registry/storage/driver/walk_test.go

```diff
@@ -45,3 +72,217 @@ func TestWalkFileRemoved(t *testing.T) {
 		t.Fatalf(err.Error())
 	}
 }
+
+func TestWalkFilesFileRemoved(t *testing.T) {
+	d := &changingFileSystem{
+		fileset: []string{"zoidberg", "bender"},
+		keptFiles: map[string]bool{
+			"zoidberg": true,
+		},
+	}
+	infos := []FileInfo{}
+	err := WalkFilesFallback(context.Background(), d, "", func(fileInfo FileInfo) error {
+		infos = append(infos, fileInfo)
+		return nil
+	})
+	if len(infos) != 1 || infos[0].Path() != "zoidberg" {
+		t.Errorf(fmt.Sprintf("unexpected path set during walk: %s", infos))
+	}
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+}
+
+func TestWalkFallback(t *testing.T) {
+	d := &fileSystem{
+		fileset: map[string][]string{
+			"/":        {"/file1", "/folder1", "/folder2"},
+			"/folder1": {"/folder1/file1"},
+			"/folder2": {"/folder2/file1"},
+		},
+	}
+	expected := []string{
+		"/file1",
+		"/folder1",
+		"/folder1/file1",
+		"/folder2",
+		"/folder2/file1",
+	}
+
+	var walked []string
+	err := WalkFallback(context.Background(), d, "/", func(fileInfo FileInfo) error {
+		if fileInfo.IsDir() != d.isDir(fileInfo.Path()) {
+			t.Fatalf("fileInfo isDir not matching file system: expected %t actual %t", d.isDir(fileInfo.Path()), fileInfo.IsDir())
+		}
+		walked = append(walked, fileInfo.Path())
+		return nil
+	})
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	compareWalked(t, expected, walked)
+}
+
+// Walk is expected to skip directory on ErrSkipDir
+func TestWalkFallbackSkipDirOnDir(t *testing.T) {
+	d := &fileSystem{
+		fileset: map[string][]string{
+			"/":        {"/file1", "/folder1", "/folder2"},
+			"/folder1": {"/folder1/file1"},
+			"/folder2": {"/folder2/file1"},
+		},
+	}
+	skipDir := "/folder1"
+	expected := []string{
+		"/file1",
+		"/folder1", // return ErrSkipDir, skip anything under /folder1
+		// skip /folder1/file1
+		"/folder2",
+		"/folder2/file1",
+	}
+
+	var walked []string
+	err := WalkFallback(context.Background(), d, "/", func(fileInfo FileInfo) error {
+		walked = append(walked, fileInfo.Path())
+		if fileInfo.Path() == skipDir {
+			return ErrSkipDir
+		}
+		if strings.Contains(fileInfo.Path(), skipDir) {
+			t.Fatalf("skipped dir %s and should not walk %s", skipDir, fileInfo.Path())
+		}
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("expected Walk to not error %v", err)
+	}
+	compareWalked(t, expected, walked)
+}
+
+func TestWalkFallbackSkipDirOnFile(t *testing.T) {
+	d := &fileSystem{
+		fileset: map[string][]string{
+			"/": {"/file1", "/file2", "/file3"},
+		},
+	}
+	skipFile := "/file2"
+	expected := []string{
+		"/file1",
+		"/file2", // return ErrSkipDir, stop early
+	}
+
+	var walked []string
+	err := WalkFallback(context.Background(), d, "/", func(fileInfo FileInfo) error {
+		walked = append(walked, fileInfo.Path())
+		if fileInfo.Path() == skipFile {
+			return ErrSkipDir
+		}
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("expected Walk to not error %v", err)
+	}
+	compareWalked(t, expected, walked)
+}
+
+// Walk is expected to skip directory on ErrSkipDir
```

Comment:
Comments for functions typically start with the function name that the comment is referring to. I don’t think we need to comment the test functions here though. Same for the other spots.